esp_idf_monitor/base/logger.py

```diff
@@ -91,16 +92,42 @@ def print(self, string, console_printer=None):  # noqa: E999
         if console_printer is None:
             console_printer = self.console.write_bytes
 
-        if self.timestamps and (self._output_enabled or self._log_file):
+        if string and self.timestamps and (self._output_enabled or self._log_file):
             t = datetime.datetime.now().strftime(self.timestamp_format)
+
             # "string" is not guaranteed to be a full line. Timestamps should be only at the beginning of lines.
             if isinstance(string, type(u'')):
                 search_patt = '\n'
-                replacement = '\n' + t + ' '
+                replacement = t + ' '
             else:
                 search_patt = b'\n'  # type: ignore
-                replacement = b'\n' + t.encode('ascii') + b' '  # type: ignore
+                replacement = t.encode('ascii') + b' '  # type: ignore
+
+            # If the output is at the start of a new line, prefix it with the timestamp replacement text.
+            if self._start_of_line:
+                string = replacement + string
+                self._start_of_line = False
+
+            # Add the search pattern to the replacement so that replace() doesn't remove the newlines.
+            replacement = search_patt + replacement
+
+            # If the new output ends with a newline, remove it so that we don't add a trailing timestamp.
+            if string.endswith(search_patt):
+                string = string[:-len(search_patt)]
+                self._start_of_line = True
```

Comment:
How about adding else and setting the `start_of_line` to false here?

Reply:
I've changed this to be the same as the `not self.timestamps` section